cloudfiles/interfaces.py

```diff
@@ -150,92 +169,139 @@ def put_file(
       raise ValueError("Compression type {} not supported.".format(compress))
 
     if (
-      content 
-      and type(content) is str 
-      and content_type 
+      content
+      and type(content) is str
+      and content_type
       and re.search('json|te?xt', content_type)
     ):
 
       content = content.encode('utf-8')
 
-    if hasattr(content, "read") and hasattr(content, "seek"):
-      with open(path, 'wb') as f:
-        shutil.copyfileobj(content, f)
-      return
+    def do_put_file():
+      if hasattr(content, "read") and hasattr(content, "seek"):
+        with open(path, 'wb') as f:
+          shutil.copyfileobj(content, f)
+        return
 
-    try:
-      with open(path, 'wb') as f:
-        f.write(content)
-    except IOError as err:
-      mkdir(os.path.dirname(path))
-      with open(path, 'wb') as f:
-        f.write(content)
+      try:
+        with open(path, 'wb') as f:
+          f.write(content)
+      except IOError as err:
+        mkdir(os.path.dirname(path))
+        with open(path, 'wb') as f:
+          f.write(content)
+
+    if self._lock_dir:
+      lock_path = os.path.join(self._lock_dir,
+                               os.path.abspath(path).replace(os.sep, "=+"))
```

Comment:
This is an interesting choice. I suspect at some point we'll run into a problem. Maybe with a file named after a hash function output. It'll probably work for now though.